### PR TITLE
fix: fix sorting problems

### LINKED ISSUE #55: Shuffled messages
Done by [#53](https://github.com/Infiziert90/ChatTwo/pull/53)

### PR DESCRIPTION
**Pull request by deansheather**
Chat history loading sorting problems:
- Refactors `Tab.Messages` to use a new type `MessageList` which handles protected access to the messages array
    - The `MessageList` API exposes write operations as methods, instead of giving callers write access to the list directly
    - Callers who want a list for reading purposes can still get a `IReadonlyList<Message>`, which will release the read lock when the scope is left (this is used by chat log rendering)
- Changes the `FetchAllTabs` call to buffer pending messages into a `List<Message>`, then apply them all at once to avoid locking contention
    - This may block the message history from getting a read lock, so the message history code has a 3ms timeout when getting a read lock. This will cause the log to blank out while messages are loaded if it takes more than 3ms to insert the messages we fetched from the database.

Incoming message sorting problems:
- Messages are now processed in order using two Queues. One Queue is for getting the content ID in the main thread (`PendingSync`) and one queue is for processing chunks and inserting into the database in a separate thread (`PendingAsync`, this one already exists).

Also fixes:
- ContentID is now fetched more reliably using a ContentIdResolver hook
- Resets `LastMessageIndex` when logging out so message content IDs can be fetched after switching characters, this broke party invites and adventure plate functionality when right clicking players

Closes #55

### CODE REVIEW

File: ChatTwo/MessageManager.cs [hunk 1]
```diff
@@ -115,16 +135,24 @@ internal void ClearAllTabs()
...
     {
         DateTimeOffset? since = null;
         if (!Plugin.Config.FilterIncludePreviousSessions)
             since = Plugin.GameStarted;
 
         var messages = Store.GetMostRecentMessages(CurrentContentId, since);
+
+        // We store the pending messages to be added to the chat log in a
+        // temporary list, and apply them all at once after filtering.
+        List<(Tab, List<Message>)> pendingTabs = Plugin.Config.Tabs.Select(tab => (tab, new List<Message>())).ToList();
```
Comment (Infiziert90) [inline ChatTwo/MessageManager.cs:148]: `var` should work fine

**Review by Infiziert90 — CHANGES_REQUESTED** [review]

File: ChatTwo/Configuration.cs [hunk 2]
```diff
@@ -244,6 +216,150 @@ internal Tab Clone()
...
+        private void PruneMaxLocked(int max)
+        {
+            while (messages.Count > max)
+            {
+                trackedMessageIds.Remove(messages[0].Id);
+                messages.RemoveAt(0);
+            }
+        }
+
+        public IEnumerator<Message> GetEnumerator()
```
[the remark this answers is not in the document]
Comment (deansheather) [reply ChatTwo/Configuration.cs:312]: Removed it since this was left over code before I made RLockedMessageList.

File: ChatTwo/MessageManager.cs [hunk 3]
```diff
@@ -21,12 +23,23 @@ internal class MessageManager : IAsyncDisposable
...
+    // After that, the message is enqueued in the PendingAsync queue, which will
+    // be consumed in a separate thread and perform more processing (emotes,
+    // URLs) as well as inserting the message into the database.
+    private Queue<PendingMessage> PendingSync { get; } = new();
+    private ConcurrentQueue<PendingMessage> PendingAsync { get; } = new();
     private readonly Thread PendingMessageThread;
     private readonly CancellationTokenSource PendingThreadCancellationToken = new();
 
+    // TODO: remove when added to client structs
+    private const string ContentIdResolverSig = "4C 8B D1 48 8B 89 ?? ?? ?? ?? 48 85 C9";
```
Comment (Infiziert90) [inline ChatTwo/MessageManager.cs:39]: You can replace this with something like <https://github.com/Infiziert90/ChatTwo/blob/main/ChatTwo/GameFunctions/Chat.cs#L108>

File: ChatTwo/MessageManager.cs [hunk 4]
```diff
@@ -44,15 +57,24 @@ internal MessageManager(Plugin plugin)
...
         Plugin.ChatGui.ChatMessageUnhandled += ChatMessage;
-        Plugin.Framework.Update += UpdateReceiver;
+        Plugin.Framework.Update += OnFrameworkUpdate;
         Plugin.ClientState.Logout += Logout;
     }
 
     public async ValueTask DisposeAsync()
     {
+        ContentIdResolverHook.Disable();
+        ContentIdResolverHook.Dispose();
```
Comment (Infiziert90) [inline ChatTwo/MessageManager.cs:75]: Dispose() is enough, it calls Disable

File: ChatTwo/MessageManager.cs [hunk 5]
```diff
@@ -80,18 +102,25 @@ private void Logout()
...
+    private void OnFrameworkUpdate(IFramework framework)
     {
         var contentId = Plugin.ClientState.LocalContentId;
         if (contentId != 0)
             LastContentId = contentId;
+
+        // Drain the PendingSync queue into the PendingAsync queue.
+        while (true)
+        {
+            if (!PendingSync.TryDequeue(out var pending)) return;
```
Comment (Infiziert90) [inline ChatTwo/MessageManager.cs:114]: newline before return

File: ChatTwo/MessageManager.cs [hunk 6]
```diff
@@ -177,31 +214,19 @@ private void ChatMessage(XivChatType type, uint senderId, SeString sender, SeStr
...
-        {
-            if (shouldGetContentId)
-                pendingMessage.ContentId = Plugin.Functions.Chat.GetContentIdForEntry(idx - 1);
-            Pending.Enqueue(pendingMessage);
-        });
+    // This hook is called immediately after receiving a message with the
+    // message's content ID. If multiple messages are received in the same tick,
+    // this will be called for each message immediately after ChatMessage is
+    // called for each message.
+    private unsafe void ContentIdResolver(RaptureLogModule* param1, ulong param2, int param3, short param4, short param5)
```
Comment (Infiziert90) [inline ChatTwo/MessageManager.cs:227]: Original must always be called, else we corrupt the game cache

So something like `ContentIdResolverHook.Original(param1, param2, param3, param4, param5);`

File: ChatTwo/Plugin.cs [hunk 7]
```diff
@@ -39,6 +41,7 @@ public sealed class Plugin : IDalamudPlugin
     [PluginService] internal static IGameConfig GameConfig { get; private set; } = null!;
     [PluginService] internal static INotificationManager Notification { get; private set; } = null!;
     [PluginService] internal static IAddonLifecycle AddonLifecycle { get; private set; } = null!;
+    [PluginService] internal static ISigScanner SigScanner { get; private set; } = null!;
```
Comment (Infiziert90) [inline ChatTwo/Plugin.cs:44]: Can be removed as we use Attribute sigs

**Review by Infiziert90 — CHANGES_REQUESTED** [review]

**Review by Infiziert90 — APPROVED** [review]

**Pull request merged**